File: bindings/csharp/Causalontology/Ed25519.cs
// Ed25519 digital signatures (RFC 8032), pure C#, BCL only.
//
// The BCL has no Ed25519, so this is a faithful port of the Python
// binding's ed25519.py onto System.Numerics.BigInteger. Slow but
// correct: intended for the conformance suite and for small tools.
// Production stores should use an optimized library; the signatures are
// byte-compatible either way (Ed25519 is deterministic, RFC 8032).
//
// CRITICAL porting note: C#'s % operator can return negatives, unlike
// Python's; every reduction goes through Mod(), which normalizes with
// ((a % p) + p) % p. The whole module is gated on the RFC 8032 TEST 1
// known answer by the conformance runner before any vector runs.

using System.Numerics;
using System.Security.Cryptography;

namespace Causalontology;

public static class Ed25519
{
    // the field prime p = 2^255 - 19
    private static readonly BigInteger P = BigInteger.Pow(2, 255) - 19;

    // the group order q = 2^252 + 27742317777372353535851937790883648493
    private static readonly BigInteger Q = BigInteger.Pow(2, 252)
        + BigInteger.Parse("27742317777372353535851937790883648493");

    // ((a % p) + p) % p — a nonnegative residue even when a is negative
    private static BigInteger Mod(BigInteger a) => ((a % P) + P) % P;

    private static BigInteger ModpInv(BigInteger x)
        => BigInteger.ModPow(Mod(x), P - 2, P);

    // the curve constant d = -121665 / 121666 (mod p)
    private static readonly BigInteger D = Mod(-121665 * ModpInv(121666));

    private static readonly BigInteger ModpSqrtM1
        = BigInteger.ModPow(2, (P - 1) / 4, P);

    // extended homogeneous coordinates (X, Y, Z, T)
    private readonly record struct Point(
        BigInteger X, BigInteger Y, BigInteger Z, BigInteger T);

    private static Point PointAdd(Point p, Point q)
    {
        var a = Mod((p.Y - p.X) * (q.Y - q.X));
        var b = Mod((p.Y + p.X) * (q.Y + q.X));
        var c = Mod(2 * p.T * q.T * D);
        var d = Mod(2 * p.Z * q.Z);
        BigInteger e = b - a, f = d - c, g = d + c, h = b + a;
        return new Point(Mod(e * f), Mod(g * h), Mod(f * g), Mod(e * h));
    }

    private static Point PointMul(BigInteger s, Point p)
    {
        var q = new Point(0, 1, 1, 0); // the neutral element
        while (s > 0)
        {
            if (!s.IsEven)
                q = PointAdd(q, p);
            p = PointAdd(p, p);
            s >>= 1;
        }
        return q;
    }

    private static bool PointEqual(Point p, Point q)
    {
        if (Mod(p.X * q.Z - q.X * p.Z) != 0)
            return false;
        if (Mod(p.Y * q.Z - q.Y * p.Z) != 0)
            return false;
        return true;
    }

    private static BigInteger? RecoverX(BigInteger y, int sign)
    {
        if (y >= P)
            return null;
        var x2 = Mod((y * y - 1) * ModpInv(D * y * y + 1));
        if (x2.IsZero)
            return sign != 0 ? null : BigInteger.Zero;
        var x = BigInteger.ModPow(x2, (P + 3) / 8, P);
        if (Mod(x * x - x2) != 0)
            x = Mod(x * ModpSqrtM1);
        if (Mod(x * x - x2) != 0)
            return null;
        if ((int)(x & 1) != sign)
            x = P - x;
        return x;
    }

    // the base point G
    private static readonly Point G = MakeBasePoint();

    private static Point MakeBasePoint()
    {
        var gy = Mod(4 * ModpInv(5));
        var gx = RecoverX(gy, 0)!.Value;
        return new Point(gx, gy, 1, Mod(gx * gy));
    }

    private static byte[] PointCompress(Point p)
    {
        var zinv = ModpInv(p.Z);
        var x = Mod(p.X * zinv);
        var y = Mod(p.Y * zinv);
        var value = y | ((x & 1) << 255);
        var bytes = new byte[32];
        value.TryWriteBytes(bytes, out _, isUnsigned: true, isBigEndian: false);
        return bytes;
    }

    private static Point? PointDecompress(byte[] s)
    {
        if (s.Length != 32)
            return null;
        var y = new BigInteger(s, isUnsigned: true, isBigEndian: false);
        var sign = (int)(y >> 255);
        y &= (BigInteger.One << 255) - 1;
        var x = RecoverX(y, sign);
        if (x is null)
            return null;
        return new Point(x.Value, y, 1, Mod(x.Value * y));
    }

    private static (BigInteger A, byte[] Prefix) SecretExpand(byte[] secret)
    {
        if (secret.Length != 32)
            throw new ArgumentException("secret key must be 32 bytes");
        var h = SHA512.HashData(secret);
        var a = new BigInteger(h.AsSpan(0, 32),
                               isUnsigned: true, isBigEndian: false);
        a &= (BigInteger.One << 254) - 8;
        a |= BigInteger.One << 254;
        return (a, h[32..]);
    }

    private static BigInteger Sha512ModQ(byte[] s)
    {
        var h = SHA512.HashData(s);
        return new BigInteger(h, isUnsigned: true, isBigEndian: false) % Q;
    }

    private static byte[] Concat(params byte[][] parts)
    {
        var total = parts.Sum(p => p.Length);
        var joined = new byte[total];
        var offset = 0;
        foreach (var part in parts)
        {
            Buffer.BlockCopy(part, 0, joined, offset, part.Length);
            offset += part.Length;
        }
        return joined;
    }

    /// <summary>The 32-byte public key for a 32-byte secret key.</summary>
    public static byte[] SecretToPublic(byte[] secret)
    {
        var (a, _) = SecretExpand(secret);
        return PointCompress(PointMul(a, G));
    }

    /// <summary>The 64-byte Ed25519 signature of msg under the 32-byte secret key.</summary>
    public static byte[] Sign(byte[] secret, byte[] msg)
    {
        var (a, prefix) = SecretExpand(secret);
        var publicKey = PointCompress(PointMul(a, G));
        var r = Sha512ModQ(Concat(prefix, msg));
        var rs = PointCompress(PointMul(r, G));
        var h = Sha512ModQ(Concat(rs, publicKey, msg));
        var s = ((r + h * a) % Q + Q) % Q;
        var sBytes = new byte[32];
        s.TryWriteBytes(sBytes, out _, isUnsigned: true, isBigEndian: false);
        return Concat(rs, sBytes);
    }

    /// <summary>True iff signature is a valid Ed25519 signature of msg under publicKey.</summary>
    public static bool Verify(byte[] publicKey, byte[] msg, byte[] signature)
    {
        if (publicKey.Length != 32 || signature.Length != 64)
            return false;
        var a = PointDecompress(publicKey);
        if (a is null)
            return false;
        var rs = signature[..32];
        var r = PointDecompress(rs);
        if (r is null)
            return false;
        var s = new BigInteger(signature.AsSpan(32),
                               isUnsigned: true, isBigEndian: false);
        if (s >= Q)
            return false;
        var h = Sha512ModQ(Concat(rs, publicKey, msg));
        var sB = PointMul(s, G);
        var hA = PointMul(h, a.Value);
        return PointEqual(sB, PointAdd(r.Value, hA));
    }
}

File: bindings/haskell/src/Causalontology/Ed25519.hs
-- | Ed25519 digital signatures (RFC 8032), pure Haskell, ported from the
-- Python binding's @ed25519.py@ over 'Integer'.
--
-- Slow but correct: intended for the conformance suite and for small
-- tools. Production stores should use an optimized library; the
-- signatures are byte-compatible either way (Ed25519 is deterministic,
-- RFC 8032).
--
-- Note on arithmetic: Haskell's @mod@ on 'Integer' is floored, so for the
-- positive moduli used here it always yields a non-negative result -
-- exactly like Python's @%@. The conformance runner asserts
-- @(-7) \`mod\` 5 == 3@ before any vector runs.
module Causalontology.Ed25519
  ( secretToPublic
  , edSign
  , edVerify
  ) where

import Causalontology.Sha2 (sha512)
import Data.Bits (shiftL, shiftR, testBit, (.&.), (.|.))
import Data.Word (Word8)

-- | The field prime 2^255 - 19.
curveP :: Integer
curveP = (2 ^ (255 :: Int)) - 19

-- | The group order.
curveQ :: Integer
curveQ = (2 ^ (252 :: Int)) + 27742317777372353535851937790883648493

-- | Square-and-multiply modular exponentiation (Python's three-argument
-- @pow@).
modPow :: Integer -> Integer -> Integer -> Integer
modPow base0 expo0 m = go (base0 `mod` m) expo0 1
  where
    go base expo acc
      | expo <= 0 = acc
      | otherwise =
          go
            ((base * base) `mod` m)
            (expo `shiftR` 1)
            (if testBit expo 0 then (acc * base) `mod` m else acc)

-- | Multiplicative inverse modulo p (Fermat).
modpInv :: Integer -> Integer
modpInv x = modPow x (curveP - 2) curveP

-- | The twisted Edwards curve constant d.
curveD :: Integer
curveD = (negate 121665 * modpInv 121666) `mod` curveP

-- | A square root of -1 modulo p.
sqrtM1 :: Integer
sqrtM1 = modPow 2 ((curveP - 1) `div` 4) curveP

-- | A point in extended homogeneous coordinates (X, Y, Z, T).
type Point = (Integer, Integer, Integer, Integer)

-- | Point addition (RFC 8032 section 5.1.4).
pointAdd :: Point -> Point -> Point
pointAdd (px, py, pz, pt) (qx, qy, qz, qt) =
  let a = ((py - px) * (qy - qx)) `mod` curveP
      b = ((py + px) * (qy + qx)) `mod` curveP
      c = (2 * pt * qt * curveD) `mod` curveP
      d = (2 * pz * qz) `mod` curveP
      e = b - a
      f = d - c
      g = d + c
      h = b + a
  in ((e * f) `mod` curveP, (g * h) `mod` curveP, (f * g) `mod` curveP, (e * h) `mod` curveP)

-- | Scalar multiplication by double-and-add.
pointMul :: Integer -> Point -> Point
pointMul scalar0 point0 = go scalar0 point0 (0, 1, 1, 0)
  where
    -- (0, 1, 1, 0) is the neutral element
    go scalar point acc
      | scalar <= 0 = acc
      | otherwise =
          go
            (scalar `shiftR` 1)
            (pointAdd point point)
            (if testBit scalar 0 then pointAdd acc point else acc)

-- | Projective equality: X1/Z1 == X2/Z2 and Y1/Z1 == Y2/Z2.
pointEqual :: Point -> Point -> Bool
pointEqual (px, py, pz, _) (qx, qy, qz, _) =
  ((px * qz - qx * pz) `mod` curveP) == 0
    && ((py * qz - qy * pz) `mod` curveP) == 0

-- | Recover the x coordinate from y and the sign bit (RFC 8032 5.1.3).
recoverX :: Integer -> Integer -> Maybe Integer
recoverX y sign
  | y >= curveP = Nothing
  | x2 == 0 = if sign /= 0 then Nothing else Just 0
  | otherwise =
      let x = modPow x2 ((curveP + 3) `div` 8) curveP
          x' =
            if ((x * x - x2) `mod` curveP) /= 0
              then (x * sqrtM1) `mod` curveP
              else x
      in if ((x' * x' - x2) `mod` curveP) /= 0
           then Nothing
           else
             if (x' .&. 1) /= sign
               then Just (curveP - x')
               else Just x'
  where
    x2 = ((y * y - 1) * modpInv ((curveD * y * y + 1) `mod` curveP)) `mod` curveP

-- | The base point's y coordinate: 4/5 mod p.
baseY :: Integer
baseY = (4 * modpInv 5) `mod` curveP

-- | The base point's x coordinate (even root).
baseX :: Integer
baseX = case recoverX baseY 0 of
  Just x -> x
  Nothing -> error "ed25519: base point recovery failed"

-- | The base point G in extended coordinates.
baseG :: Point
baseG = (baseX, baseY, 1, (baseX * baseY) `mod` curveP)

-- | Compress a point to 32 little-endian bytes (y with the x sign bit).
pointCompress :: Point -> [Word8]
pointCompress (px, py, pz, _) =
  let zinv = modpInv pz
      x = (px * zinv) `mod` curveP
      y = (py * zinv) `mod` curveP
  in intToLe 32 (y .|. ((x .&. 1) `shiftL` 255))

-- | Decompress 32 bytes to a point; 'Nothing' if not on the curve.
pointDecompress :: [Word8] -> Maybe Point
pointDecompress bytes
  | length bytes /= 32 = Nothing
  | otherwise =
      let raw = leToInt bytes
          sign = raw `shiftR` 255
          y = raw .&. ((1 `shiftL` 255) - 1)
      in case recoverX y sign of
           Nothing -> Nothing
           Just x -> Just (x, y, 1, (x * y) `mod` curveP)

-- | Little-endian encoding of a non-negative integer into n bytes.
intToLe :: Int -> Integer -> [Word8]
intToLe n v = [ fromIntegral ((v `shiftR` (8 * i)) .&. 0xFF) | i <- [0 .. n - 1] ]

-- | Little-endian decoding of a byte string.
leToInt :: [Word8] -> Integer
leToInt = foldr (\b acc -> (acc `shiftL` 8) .|. fromIntegral b) 0

-- | Expand a 32-byte secret into the clamped scalar and the prefix.
secretExpand :: [Word8] -> (Integer, [Word8])
secretExpand secret
  | length secret /= 32 = error "ed25519: secret key must be 32 bytes"
  | otherwise =
      let h = sha512 secret
          a0 = leToInt (take 32 h)
          a1 = a0 .&. ((1 `shiftL` 254) - 8)
          a2 = a1 .|. (1 `shiftL` 254)
      in (a2, drop 32 h)

-- | SHA-512 reduced modulo the group order.
sha512ModQ :: [Word8] -> Integer
sha512ModQ bytes = leToInt (sha512 bytes) `mod` curveQ

-- | The 32-byte public key for a 32-byte secret key.
secretToPublic :: [Word8] -> [Word8]
secretToPublic secret = pointCompress (pointMul a baseG)
  where
    (a, _) = secretExpand secret

-- | The 64-byte Ed25519 signature of a message under a 32-byte secret key.
edSign :: [Word8] -> [Word8] -> [Word8]
edSign secret msg =
  let (a, prefix) = secretExpand secret
      publicA = pointCompress (pointMul a baseG)
      r = sha512ModQ (prefix ++ msg)
      rBytes = pointCompress (pointMul r baseG)
      h = sha512ModQ (rBytes ++ publicA ++ msg)
      s = (r + h * a) `mod` curveQ
  in rBytes ++ intToLe 32 s

-- | True iff the signature is a valid Ed25519 signature of the message
-- under the public key.
edVerify :: [Word8] -> [Word8] -> [Word8] -> Bool
edVerify public msg signature
  | length public /= 32 || length signature /= 64 = False
  | otherwise =
      let rBytes = take 32 signature
      in case (pointDecompress public, pointDecompress rBytes) of
           (Just aPoint, Just rPoint) ->
             let s = leToInt (drop 32 signature)
             in (s < curveQ)
                  && let h = sha512ModQ (rBytes ++ public ++ msg)
                         sB = pointMul s baseG
                         hA = pointMul h aPoint
                     in pointEqual sB (pointAdd rPoint hA)
           _ -> False
